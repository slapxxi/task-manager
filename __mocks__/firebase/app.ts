import mock from '.';

export default mock;
