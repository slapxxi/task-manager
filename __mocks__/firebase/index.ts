import { set } from 'lodash';

Date.now = () => 69;

let databaseMock = {
  tasks: {
    'first-task': {
      title: 'first',
      deadline: 0,
      createdAt: 1,
    },
    'second-task': {
      title: 'second',
      deadline: 2668,
      createdAt: 2,
    },
    'third-task': {
      title: 'third',
      tags: ['html'],
      deadline: 0,
      project: 'test',
    },
  },
  tags: {
    html: {
      name: 'HTML',
    },
    css: {
      name: 'CSS',
    },
  },
  projects: {
    test: {
      name: 'Test Project',
    },
  },
};

const snapshot = {
  val() {
    return databaseMock;
  },
  tasks: {
    val() {
      return databaseMock.tasks;
    },
  },
  tags: {
    val() {
      return databaseMock.tags;
    },
  },
};

const mock = {
  __reset() {
    databaseMock = {
      tasks: {
        'first-task': {
          title: 'first',
          deadline: 0,
          createdAt: 1,
        },
        'second-task': {
          title: 'second',
          deadline: 2668,
          createdAt: 2,
        },
        'third-task': {
          title: 'third',
          tags: ['html'],
          deadline: 0,
          project: 'test',
        },
      },
      tags: {
        html: {
          name: 'HTML',
        },
        css: {
          name: 'CSS',
        },
      },
      projects: {
        test: {
          name: 'Test Project',
        },
      },
    };
  },
  database() {
    return {
      ref(path: string) {
        if (path === '/') {
          return {
            on(event, fn) {
              fn(snapshot);
              const interval = setInterval(() => {
                fn(snapshot);
              }, 1);
              return interval;
            },
            off(event, fn) {
              clearInterval(fn);
            },
          };
        }
        return {
          on(event, fn) {
            fn(snapshot.tasks);
            setInterval(() => {
              fn(snapshot.tasks);
            }, 1);
          },
          once(event) {
            return new Promise((resolve) => resolve(snapshot));
          },
          set(value) {
            const objectPath = path.replace(/\//g, '.').slice(1);
            set(databaseMock, objectPath, value);
          },
        };
      },
    };
  },
};

export default mock;
