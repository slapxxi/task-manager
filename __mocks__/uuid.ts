const uuidMock = {
  counter: 0,
  __reset() {
    this.counter = 0;
  },
  v4() {
    const id = `unique-id-${this.counter}`;
    this.counter += 1;
    return id;
  },
};

export default uuidMock;
