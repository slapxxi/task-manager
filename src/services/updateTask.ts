import { createTask } from '@lib';
import { DBTask, Tag, Task } from '@local/types';
import firebase from 'firebase';
import { find } from 'lodash';
import uuid from 'uuid';
import createDBTag from './createDBTag';

function updateTask(task: Task, tags: Tag[]) {
  const id = task.id || uuid.v4();
  if (task.tags) {
    task = updateWithTags(task, tags);
  }
  firebase
    .database()
    .ref(`/tasks/${id}`)
    .set(taskToDBTask(task));
}

function updateWithTags(task: Task, tags: Tag[]) {
  return {
    ...task,
    tags: matchTags(task.tags as Tag[], tags),
  };
}

function matchTags(candidateTags: Tag[], existingTags: Tag[]) {
  return candidateTags.reduce<Tag[]>(
    (acc, current) => [
      ...acc,
      find(
        existingTags,
        (t) =>
          (current.name &&
            t.name.toLowerCase() === current.name.toLowerCase()) ||
          t.id === current.id,
      ) || {
        id: createDBTag({ name: current.name }),
        name: current.name,
      },
    ],
    [],
  );
}

function taskToDBTask(params: Task): DBTask {
  const result = {
    ...createTask(params),
    tags: (params.tags as Tag[]).map((t) => t.id),
  };
  delete result.id;
  return result;
}

export default updateTask;
