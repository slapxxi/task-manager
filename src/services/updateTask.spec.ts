import { createTag, createTask } from '@lib';
import firebase from 'firebase/app';
import uuid from 'uuid';
import updateTask from './updateTask';

const tags = [
  createTag({ id: 'html', name: 'HTML' }),
  createTag({ id: 'css', name: 'CSS' }),
];

beforeEach(async () => {
  // @ts-ignore
  firebase.__reset();
  // @ts-ignore
  uuid.__reset();
});

it('updates task', async () => {
  updateTask(createTask({ title: 'task' }), tags);
  const result = await firebase
    .database()
    .ref('/tasks')
    .once('value');
  expect(result.val()).toMatchSnapshot();
});

it('reuses existing tags', async () => {
  updateTask(
    createTask({
      title: 'task',
      tags: [{ name: 'css' }, { name: 'html' }, { name: 'ruby' }],
    }),
    tags,
  );
  const result = await firebase
    .database()
    .ref('/tasks')
    .once('value');
  expect(result.val()).toMatchSnapshot();
});
