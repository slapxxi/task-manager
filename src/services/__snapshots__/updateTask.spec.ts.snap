// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reuses existing tags 1`] = `
Object {
  "projects": Object {
    "test": Object {
      "name": "Test Project",
    },
  },
  "tags": Object {
    "css": Object {
      "name": "CSS",
    },
    "html": Object {
      "name": "HTML",
    },
    "unique-id-1": Object {
      "name": "ruby",
    },
  },
  "tasks": Object {
    "first-task": Object {
      "createdAt": 1,
      "deadline": 0,
      "title": "first",
    },
    "second-task": Object {
      "createdAt": 2,
      "deadline": 2668,
      "title": "second",
    },
    "third-task": Object {
      "deadline": 0,
      "project": "test",
      "tags": Array [
        "html",
      ],
      "title": "third",
    },
    "unique-id-0": Object {
      "completed": false,
      "createdAt": 1970-01-01T00:00:00.069Z,
      "deadline": 0,
      "description": "",
      "project": "",
      "subtasks": Array [],
      "tags": Array [
        "css",
        "html",
        "unique-id-1",
      ],
      "title": "task",
    },
  },
}
`;

exports[`updates task 1`] = `
Object {
  "projects": Object {
    "test": Object {
      "name": "Test Project",
    },
  },
  "tags": Object {
    "css": Object {
      "name": "CSS",
    },
    "html": Object {
      "name": "HTML",
    },
  },
  "tasks": Object {
    "first-task": Object {
      "createdAt": 1,
      "deadline": 0,
      "title": "first",
    },
    "second-task": Object {
      "createdAt": 2,
      "deadline": 2668,
      "title": "second",
    },
    "third-task": Object {
      "deadline": 0,
      "project": "test",
      "tags": Array [
        "html",
      ],
      "title": "third",
    },
    "unique-id-0": Object {
      "completed": false,
      "createdAt": 1970-01-01T00:00:00.069Z,
      "deadline": 0,
      "description": "",
      "project": "",
      "subtasks": Array [],
      "tags": Array [],
      "title": "task",
    },
  },
}
`;
